Framework/Kernel/src/TimeROI.cpp

```diff
@@ -0,0 +1,267 @@
+void assert_increasing(const DateAndTime &startTime, const DateAndTime &stopTime) {
+  if (!bool(startTime < stopTime)) {
+    std::stringstream msg;
+    msg << startTime << " and  " << stopTime << " are not in increasing order";
+    throw std::runtime_error(msg.str());
+  }
+}
+
+} // namespace
+
+TimeROI::TimeROI() : m_roi{NAME} {}
+
+TimeROI::TimeROI(const Types::Core::DateAndTime &startTime, const Types::Core::DateAndTime &stopTime) : m_roi{NAME} {
+  this->addROI(startTime, stopTime);
+}
+
+void TimeROI::addROI(const std::string &startTime, const std::string &stopTime) {
+  this->addROI(DateAndTime(startTime), DateAndTime(stopTime));
+}
+
+void TimeROI::addROI(const Types::Core::DateAndTime &startTime, const Types::Core::DateAndTime &stopTime) {
+  assert_increasing(startTime, stopTime);
+  m_roi.addValue(startTime, ROI_USE);
+  m_roi.addValue(stopTime, ROI_IGNORE);
+}
+
+void TimeROI::addROI(const std::time_t &startTime, const std::time_t &stopTime) {
+  this->addROI(DateAndTime(startTime), DateAndTime(stopTime));
+}
+
+void TimeROI::addMask(const std::string &startTime, const std::string &stopTime) {
+  this->addMask(DateAndTime(startTime), DateAndTime(stopTime));
+}
+
+void TimeROI::addMask(const Types::Core::DateAndTime &startTime, const Types::Core::DateAndTime &stopTime) {
+  assert_increasing(startTime, stopTime);
+  m_roi.addValue(startTime, ROI_IGNORE);
+  m_roi.addValue(stopTime, ROI_USE);
+}
+
+void TimeROI::addMask(const std::time_t &startTime, const std::time_t &stopTime) {
+  this->addMask(DateAndTime(startTime), DateAndTime(stopTime));
+}
+
+bool TimeROI::valueAtTime(const DateAndTime &time) const {
+  if (this->empty() || time < m_roi.firstTime()) {
+    return ROI_IGNORE;
+  } else {
+    return m_roi.getSingleValue(time);
+  }
+}
+
+/// get a list of all unique times. order is not guaranteed
+std::vector<DateAndTime> TimeROI::getAllTimes(const TimeROI &other) {
+
+  std::set<DateAndTime> times_set;
+  const auto times_lft = this->m_roi.timesAsVector();
+  for (const auto time : times_lft)
+    times_set.insert(time);
+  const auto times_rgt = other.m_roi.timesAsVector();
+  for (const auto time : times_rgt)
+    times_set.insert(time);
+
+  // copy into the vector
+  std::vector<DateAndTime> times_all;
+  times_all.assign(times_set.begin(), times_set.end());
+
+  return times_all;
+}
+
+void TimeROI::replaceValues(const std::vector<DateAndTime> &times, const std::vector<bool> &values) {
+  if (times.size() != values.size()) {
+    std::stringstream msg;
+    msg << "Times and Values are different size: " << times.size() << " != " << values.size();
+    throw std::runtime_error(msg.str());
+  }
+
+  // remove all current values
+  this->m_roi.clear();
+
+  // see if everything to add is "IGNORE"
+  bool set_values = std::any_of(values.cbegin(), values.cend(), [](bool value) { return value; });
+
+  // set the values if there are any use regions
+  if (set_values) {
```

Comment:
Aren't the values supposed to alternate `ROI_USE, ROI_IGNORE, ...` for the final list to be valid? Validation is required

Reply:
They don't have to be, but with the updated behavior in `addROI` and `addMask` they tend to be. Note that this is a private method. The main exception is when this is called by `update_intersection` and `update_union` which evaluate both input `TimeROI` at all possible times that either list which creates duplicate/redundant values. Since `removeRedundantEntries` is moderately expensive, `replaceValues` doesn't automatically call it.